### PR TITLE
Straight-through estimator for nvfp4

### PR DESCRIPTION
**Pull request by wenscarl**
For NVFP4, the global scaling factors are either calibrated or provided based on history. If the scaling factor is clipped, zero out the corresponding gradients; otherwise, pass the gradients through unchanged.

### CODE REVIEW

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 1]
```diff
@@ -489,18 +489,24 @@ def quantize(x, config):
...
-    scales_q = jax.lax.optimization_barrier(scales.astype(jnp.float8_e4m3fn))
-    scaled_x = x / (scales_q.astype(jnp.float32) *
-                    config.global_scale).astype(x.dtype)
+    SCALE_MAX = jnp.finfo(config.scale_type).max.astype(x.dtype)
+
+    prev_amax = config.global_scale * (MAX * SCALE_MAX)
+    scales_q = jnp.clip(
+        (amax / prev_amax) * SCALE_MAX, 0, SCALE_MAX
+    ).astype(config.scale_type)
+    x_q = jnp.where(
```
Comment (kaixih) [inline jax/_src/cudnn/scaled_matmul_stablehlo.py:503]: This is not right. We don't need to change to use where explicitly.

Comment (kaixih) [reply jax/_src/cudnn/scaled_matmul_stablehlo.py:503]: The original ```clipped_x = jnp.clip(scaled_x, -MAX, MAX)
  x_q = clipped_x.astype(config.data_type)``` can already do this logic.

File: tests/scaled_matmul_stablehlo_test.py [hunk 2]
```diff
@@ -194,6 +194,11 @@ def generate_nvfp4_quantized_tensors(dot_config, output_type):
   amax_a = jnp.max(jnp.abs(a)).astype(jnp.float32)
   amax_b = jnp.max(jnp.abs(b)).astype(jnp.float32)
 
+  # To emulate calibrated amax
+  amax_sf = 0.9
```
Comment (kaixih) [inline tests/scaled_matmul_stablehlo_test.py:198]: can we add a parameterized test like 
```
def test***(enable_grad_clip, ****):
...
# We use a smaller amax to test if the gradients for the clipped values are zeroed out.
amax_factor = 0.9 if enable_grad_clip else 1.0
```

Comment (kaixih) [reply tests/scaled_matmul_stablehlo_test.py:198]: And we can skip it for the mxfp8 tests.

File: tests/scaled_matmul_stablehlo_test.py [hunk 3]
```diff
@@ -567,8 +572,27 @@ def fwd(a, b, is_ref=False, use_normalized=False):
...
+      data_max = jnp.finfo(jnp.float4_e2m1fn).max.astype(output_type)
+      scale_max = jnp.finfo(jnp.float8_e4m3fn).max.astype(output_type)
+      prev_amax_a = a_gs * data_max * scale_max
+      prev_amax_b = b_gs * data_max * scale_max
+
+      x_grad_ref = _grad_clip(prev_amax_a, a_raw, x_grad_ref)
+      w_grad_ref = _grad_clip(prev_amax_b, b_raw, w_grad_ref)
       self.assertArraysAllClose(x_grad, x_grad_ref, rtol=1e-2, atol=1e1)
       self.assertArraysAllClose(w_grad, w_grad_ref, rtol=1e-2, atol=1e1)
+      # Verify straight_through_estimator
```
Comment (kaixih) [inline tests/scaled_matmul_stablehlo_test.py:587]: I think the above is already check the zeros in the outputs, since the ref results already contain the zeros. Why do we need to check it again below?

Comment (wenscarl) [reply tests/scaled_matmul_stablehlo_test.py:587]: Removed.

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 4]
```diff
@@ -492,13 +492,17 @@ def quantize(x, config):
...
     assert config.global_scale.dtype == jnp.float32
+    SCALE_MAX = jnp.finfo(config.scale_type).max.astype(x.dtype)
 
-    scales = scales / config.global_scale
-    scales_q = jax.lax.optimization_barrier(scales.astype(jnp.float8_e4m3fn))
-    scaled_x = x / (scales_q.astype(jnp.float32) *
-                    config.global_scale).astype(x.dtype)
+    prev_amax = config.global_scale * (MAX * SCALE_MAX)
+    scales_q = jnp.clip(
+        (amax / prev_amax) * SCALE_MAX, 0, SCALE_MAX
```
Comment (kaixih) [inline jax/_src/cudnn/scaled_matmul_stablehlo.py:499]: Are we sure this is correct? Shouldn't it be clip to `-SCALE_MAX` to `SCALE_MAX`?

Also, can we keep using the formula of `scales / config.global_scale` instead of `(amax / prev_amax) * SCALE_MAX`? This would be consistent with how we scale any target values?

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 5]
```diff
@@ -492,13 +492,17 @@ def quantize(x, config):
...
+    )
+    scaled_x = x / scales_q
+    scales_q = scales_q.astype(config.scale_type)
```
Comment (kaixih) [inline jax/_src/cudnn/scaled_matmul_stablehlo.py:502]: I don't think we should change the order and we should use the fp8 scales during the scaling of the x, meaning it has to be scales->fp8->cvt->fp32 for the scales.

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 6]
```diff
@@ -639,6 +643,14 @@ def scaled_dot_bwd(dimension_numbers, preferred_element_type, configs, res, g):
   }
   grad_lhs = scaled_dot_general_transpose_lhs(*args, **lhs_kw_args)
   grad_rhs = scaled_dot_general_transpose_rhs(*args, **rhs_kw_args)
+
+  if configs[2].mode == "nvfp4":
```
Comment (kaixih) [inline jax/_src/cudnn/scaled_matmul_stablehlo.py:647]: Can we add some comments here? like
```
We apply a Straight-Through Estimator (STE) with zero-out behavior: if inputs are clipped during quantization in fprop, their corresponding gradients are zeroed out; otherwise, they pass through unchanged.
```

File: jax/_src/cudnn/scaled_matmul_stablehlo.py (hunk 4 above)
Comment (wenscarl) [reply jax/_src/cudnn/scaled_matmul_stablehlo.py:499]: scales are non-negative value so [0, MAX] should be range.

**Review by mattjj — APPROVED** [review]
Thanks!

**Comment by mattjj** [comment]
I suspect the doc test failure was an unrelated flake.
